### PR TITLE
runtime: support setting the timeout of guest OCI hooks

### LINKED ISSUE #4856: Support setting the timeout of guest OCI hooks
**Is your feature request related to a problem? Please describe.**
 in my case, there is a prestart-hook need to be called before the user-specified command is executed, but the prestart-hook takes a long time(more than 10s) and causes the kata container creates fail. 
command like:
```
$ ctr --debug run --runtime "io.containerd.kata.v2" --label cdi.k8s.io/nvidia_devices0="nvidia.com/gpu=gpu0" --rm -t "docker.io/nvidia/cuda:11.6.0-base-ubuntu20.04" cuda-1 bash
ctr: failed to create shim task: ETIMEDOUT: Connection timed out: unknown
``` 
log:
```
Aug 09 15:46:50 ai-2080ti-27 kata[147540]: time="2022-08-09T15:46:50.001495438+08:00" level=debug msg="reading guest console" console-protocol=unix console-url=/run/vc/vm/cuda-1/console.sock name=containerd-shim-v2 pid=147540 sandbox=cuda-1 source=virtcontainers subsystem=sandbox vmconsole="{\"msg\":\"prestart hook\",\"level\":\"INFO\",\"ts\":\"2022-08-09T07:46:49.979585542Z\",\"source\":\"agent\",\"cid\":\"cuda-1\",\"module\":\"rustjail\",\"name\":\"kata-agent\",\"version\":\"0.1.0\",\"action\":\"join-namespaces\",\"pid\":\"80\",\"subsystem\":\"container\",\"eid\":\"cuda-1\"}"
Aug 09 15:46:50 ai-2080ti-27 kata[147540]: time="2022-08-09T15:46:50.14299059+08:00" level=debug msg="reading guest console" console-protocol=unix console-url=/run/vc/vm/cuda-1/console.sock name=containerd-shim-v2 pid=147540 sandbox=cuda-1 source=virtcontainers subsystem=sandbox vmconsole="{\"msg\":\"got uevent message\",\"level\":\"INFO\",\"ts\":\"2022-08-09T07:46:50.120845002Z\",\"version\":\"0.1.0\",\"source\":\"agent\",\"subsystem\":\"uevent\",\"name\":\"kata-agent\",\"pid\":\"80\",\"event\":\"Uevent { action: \\\"add\\\", devpath: \\\"/bus/pci/drivers/nvidia-nvswitch\\\", devname: \\\"\\\", subsystem: \\\"drivers\\\", seqnum: \\\"999\\\", interface: \\\"\\\" }\"}"
Aug 09 15:46:50 ai-2080ti-27 kata[147540]: time="2022-08-09T15:46:50.14485645+08:00" level=debug msg="reading guest console" console-protocol=unix console-url=/run/vc/vm/cuda-1/console.sock name=containerd-shim-v2 pid=147540 sandbox=cuda-1 source=virtcontainers subsystem=sandbox vmconsole="[    6.682603] "
Aug 09 15:47:00 ai-2080ti-27 kata[147540]: time="2022-08-09T15:47:00.003865742+08:00" level=debug msg="reading guest console" console-protocol=unix console-url=/run/vc/vm/cuda-1/console.sock name=containerd-shim-v2 pid=147540 sandbox=cuda-1 source=virtcontainers subsystem=sandbox vmconsole="{\"msg\":\"create container process error ETIMEDOUT: Connection timed out\",\"level\":\"ERRO\",\"ts\":\"2022-08-09T07:46:59.980725078Z\",\"subsystem\":\"container\",\"module\":\"rustjail\",\"version\":\"0.1.0\",\"name\":\"kata-agent\",\"eid\":\"cuda-1\",\"pid\":\"80\",\"cid\":\"cuda-1\",\"source\":\"agent\"}"
Aug 09 15:47:00 ai-2080ti-27 kata[147540]: time="2022-08-09T15:47:00.004001615+08:00" level=error msg="createContainer failed" error="rpc error: code = Internal desc = ETIMEDOUT: Connection timed out" name=containerd-shim-v2 pid=147540 sandbox=cuda-1 source=virtcontainers subsystem=kata_agent
```
AFAIK, the default timeout for hooks is 10s, ref: https://github.com/kata-containers/kata-containers/blob/main/src/agent/rustjail/src/container.rs#L1553

I think we need to provide a way to support users to modify the timeout(if it already exists, pls tell me, thanks). 

**Describe the solution you'd like**

get the user-defined timeout from the kata config, and pass it to the agent by create sandbox request.
I can provide some help if needed.

### LINKED ISSUE #4865: When I use 'kata-runtime check' command on node using hygon cpu, it prompts "unknow cpu type"
When I use 'kata-runtime check' command on node using hygon cpu, it prompts "unknow cpu type"

图片

### PR DESCRIPTION
**Pull request by soulseen**
currently, the default timeout of guest OCI hooks is 10s, and no way to change it. This commit aims to support modifying the timeout of guest OCI hooks.

fix #4856 

Signed-off-by: soulsenn <zhuxiaoyang1996@gmail.com>

### CODE REVIEW

**Comment by soulseen** [comment]
@liubin PTAL, thanks.

File: src/agent/src/sandbox.rs [hunk 1]
```diff
@@ -341,6 +346,7 @@ impl Sandbox {
                     .unwrap()
                     .to_owned(),
                 args: vec![name.to_str().unwrap().to_owned(), hook_type.to_owned()],
+                timeout,
```
Comment (soulseen) [inline src/agent/src/sandbox.rs:349]: consistent with before., the `timeout` will limit each hook separately.

**Comment by liubin** [comment]
This PR modified `src/libs/protocols/protos/agent.proto`, that is, `runtime-rs` should be updated too, and also the `src/libs/kata-types` to keep the synchronization between all components.

Update agent protocol may be a heavy work, should we pass some additional config info through annotation?

/cc @kata-containers/architecture-committee 

@soulseen, from the user perspective, they can only set timeout through annotation? I think it should also give user the change to set it by modifying the configuration file to do a node level setup, like:

https://github.com/kata-containers/kata-containers/blob/33360f1710451293d7fcce301de02a16a1000139/src/runtime/config/configuration-qemu.toml.in#L363

**Comment by soulseen** [comment]
> @soulseen, from the user perspective, they can only set timeout through annotation? I think it should also give user the change to set it by modifying the configuration file to do a node level setup, like:
> 

@liubin, Of course they can, but I forgot to describe it in the default configuration file. Already updated.

**Comment by soulseen** [comment]
@bergwolf @fidencio @amshinde @egernst PTAL, thanks a lot.

File: src/runtime/pkg/katatestutils/utils.go [hunk 2]
```diff
@@ -318,6 +319,7 @@ func MakeRuntimeConfigFileData(config RuntimeConfigOptions) string {
 	msize_9p = ` + strconv.FormatUint(uint64(config.DefaultMsize9p), 10) + `
 	enable_debug = ` + strconv.FormatBool(config.HypervisorDebug) + `
 	guest_hook_path = "` + config.DefaultGuestHookPath + `"
+    guest_hook_timeout = ` + strconv.FormatUint(uint64(config.DefaultGuestHookTimeout), 10) + `
```
Comment (amshinde) [inline src/runtime/pkg/katatestutils/utils.go:322]: Please fix the formatting here with gofmt.

File: src/runtime/pkg/katautils/config-settings.go.in [hunk 3]
```diff
@@ -79,6 +79,7 @@ const defaultHotplugVFIOOnRootBus bool = false
 const defaultPCIeRootPort = 0
 const defaultEntropySource = "/dev/urandom"
 const defaultGuestHookPath string = ""
+const defaultGuestHookTimeout int32 = 10
```
Comment (amshinde) [inline src/runtime/pkg/katautils/config-settings.go.in:82]: nit:  I would make this as unsigned value.

File: .gitignore [hunk 4]
```diff
@@ -11,4 +11,4 @@ src/agent/protocols/src/*.rs
 !src/agent/protocols/src/lib.rs
 build
 src/tools/log-parser/kata-log-parser
-
+.idea/
```
Comment (amshinde) [inline .gitignore:14]: If you intend on adding this, please add this as a separate commit as it is unrelated.

File: src/agent/src/sandbox.rs [hunk 5]
```diff
@@ -788,7 +794,8 @@ mod tests {
         assert!(file.set_permissions(perm).is_ok());
         assert!(File::create(tmpdir.path().join("poststop").join("poststop.sh")).is_ok());
 
-        assert!(s.add_hooks(tmpdir_path).is_ok());
+        let timeout: Option<i32> = std::option::Option::Some(10);
```
Comment (amshinde) [inline src/agent/src/sandbox.rs:797]: Please check the formatting here.

**Comment by amshinde** [comment]
@soulseen I have left a few comments. I had a question regarding the intent about increasing the timeout value. Afaik, the container runtime such as containerd/docker or k8s itself may impose timeouts on the container runtime returning a response, increasing the prestart hook timeout may trigger timeouts up the stack.

cc @liubin

File: src/runtime/pkg/katatestutils/utils.go (hunk 2 above)
Comment (soulseen) [reply src/runtime/pkg/katatestutils/utils.go:322]: It was formatted locally, but I'm not sure why it shows differently in Github.
image

File: src/runtime/pkg/katautils/config-settings.go.in (hunk 3 above)
Comment (soulseen) [reply src/runtime/pkg/katautils/config-settings.go.in:82]: but in OCI defined, it also used `i32`, I think we should keep it consistent.
https://github.com/kata-containers/kata-containers/blob/main/src/libs/oci/src/lib.rs#L188

**Comment by soulseen** [comment]
> @soulseen I have left a few comments. I had a question regarding the intent about increasing the timeout value. Afaik, the container runtime such as containerd/docker or k8s itself may impose timeouts on the container runtime returning a response, increasing the prestart hook timeout may trigger timeouts up the stack.
> 
> cc @liubin

@amshinde I have not found the timeout of containerd/docker or k8s. And in my test with contained and k8s, It works fine even though the prestart hook takes 20s.

**Comment by soulseen** [comment]
@amshinde @liubin  hi, I found there is a params `--runtime-request-timeout` of `kubelet` can set runtime requests timeout:
```
--runtime-request-timeout duration                         Timeout of all runtime requests except long running request - pull, logs, exec and attach. When timeout exceeded, kubelet will cancel the request, throw out an error and retry later. (default 2m0s) (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
```

and I think the process of creating container is async, `kubelet` will sync pod state after contained(kata) is returned.(if I wrong pls correct me)

**Comment by soulseen** [comment]
@amshinde PTAL

File: src/libs/protocols/protos/agent.proto [hunk 6]
```diff
@@ -293,8 +293,10 @@ message CreateSandboxRequest {
 	// This field, if non-empty, designates an absolute path to a directory
 	// that the agent will search for OCI hooks to run within the guest.
 	string guest_hook_path = 6;
+	// GuestHookTimeout is the number of seconds before aborting the hook, If set, it must be greater than zero.
+	int32 guest_hook_timeout = 7;
```
Comment (bergwolf) [inline src/libs/protocols/protos/agent.proto:297]: Please just add new fields at the end of a struct. Inserting things in the middle will break grpc compatibility.

**Review by bergwolf — COMMENTED** [review]
Just one comment, otherwise lgtm. Thanks @soulseen ! This is definitely useful.

File: src/runtime/pkg/katatestutils/utils.go [hunk 7]
```diff
@@ -320,6 +321,7 @@ func MakeRuntimeConfigFileData(config RuntimeConfigOptions) string {
...
```
Comment (bergwolf) [inline src/runtime/pkg/katatestutils/utils.go:324]: please go fmt the code.

File: src/runtime/config/configuration-acrn.toml.in [hunk 8]
```diff
@@ -117,7 +117,12 @@ block_device_driver = "@DEFBLOCKSTORAGEDRIVER_ACRN@"
 # Warnings will be logged if any error is encountered while scanning for hooks,
 # but it will not abort container execution.
 #guest_hook_path = "/usr/share/oci/hooks"
-
+#
+# Hook timeout is the number of seconds before aborting the hook.
+# If use the OCI Hook, user can set a hook timeout to ensure that the hook has enough time to execute.
+# If set, it must be greater than zero. default value is 10s
```
Comment (bergwolf) [inline src/runtime/config/configuration-acrn.toml.in:123]: 1. The guest hook is a kata-defined hook. It is not an OCI hook.
2. We should allow zero timeout value for those who aren't sure about how long their hooks are expected to execute.

I'd suggest something like this:
```
Guest hook timeout is the number of seconds to wait for a hook execution before aborting it.
It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
```

File: src/runtime/pkg/katatestutils/utils.go (hunk 7 above)
Comment (soulseen) [reply src/runtime/pkg/katatestutils/utils.go:324]: This is already the result of a `go fmt`, but it's strange why GitHub shows it like this.

File: src/runtime/config/configuration-acrn.toml.in (hunk 8 above)
Comment (soulseen) [reply src/runtime/config/configuration-acrn.toml.in]: @bergwolf done

File: src/agent/src/rpc.rs [hunk 9]
```diff
@@ -1218,7 +1218,9 @@ impl agent_ttrpc::AgentService for AgentService {
             s.running = true;
 
             if !req.guest_hook_path.is_empty() {
-                let _ = s.add_hooks(&req.guest_hook_path).map_err(|e| {
+                let timeout: Option<i32> = Some(req.guest_hook_timeout);
```
Comment (bergwolf) [inline src/agent/src/rpc.rs:1221]: As I mentioned in the earlier comment, it should accept zero timeouts in case admins aren't sure about how long the hook will execute. It seems that the configuration comment is changed but the actual logic is not implemented yet?

File: src/libs/protocols/protos/agent.proto [hunk 10]
```diff
@@ -295,6 +295,8 @@ message CreateSandboxRequest {
 	string guest_hook_path = 6;
 	// This field is the list of kernel modules to be loaded in the guest kernel.
 	repeated KernelModule kernel_modules = 7;
+	// GuestHookTimeout is the number of seconds before aborting the hook, If set, it must be greater than zero.
```
Comment (bergwolf) [inline src/libs/protocols/protos/agent.proto:298]: The comment need to be changed as well.

**Comment by Ji-Xinyou** [comment]
Your commits need a rebase.

File: src/agent/src/rpc.rs (hunk 9 above)
Comment (Ji-Xinyou) [reply src/agent/src/rpc.rs:1221]: Previously the default timeout is 10 sec. I would suggest the 0 sec's semantic could be blocking. WDYT?

Comment (bergwolf) [reply src/agent/src/rpc.rs:1221]: sgtm!

Comment (Ji-Xinyou) [reply src/agent/src/rpc.rs:1221]: \cc @soulseen , PTAL.

Comment (soulseen) [reply src/agent/src/rpc.rs:1212]: @Ji-Xinyou @bergwolf agree and updated, please take a review. thanks!

**Comment by soulseen** [comment]
> Your commits need a rebase.

done

File: src/agent/rustjail/src/container.rs [hunk 11]
```diff
@@ -1604,7 +1604,7 @@ pub async fn execute_hook(logger: &Logger, h: &Hook, st: &OCIState) -> Result<()
 
     // if timeout is set if hook, then use the specified value
```
Comment (Ji-Xinyou) [inline src/agent/rustjail/src/container.rs:1605]: IMO, you may add the semantic of zero-timeout through some comments here as a hint for `developers`, since this is somewhat manually assigned.

File: src/libs/protocols/protos/agent.proto [hunk 12]
```diff
@@ -295,6 +295,9 @@ message CreateSandboxRequest {
 	string guest_hook_path = 6;
 	// This field is the list of kernel modules to be loaded in the guest kernel.
 	repeated KernelModule kernel_modules = 7;
+	// This fieldis the number of seconds to wait for a hook execution before aborting it.
+	// It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
```
Comment (Ji-Xinyou) [inline src/libs/protocols/protos/agent.proto:299]: - The first line: you need to add a `whitespace` between `field` and `is`

File: src/libs/protocols/protos/agent.proto [hunk 13]
```diff
@@ -295,6 +295,9 @@ message CreateSandboxRequest {
...
+	int32 guest_hook_timeout = 8;
```
Comment (Ji-Xinyou) [inline src/libs/protocols/protos/agent.proto:300]: nit: If the guest hook is a `non-negative` integer, why we are using `int32` here?

File: src/runtime/pkg/katautils/config.go [hunk 14]
```diff
@@ -521,6 +522,13 @@ func (h hypervisor) guestHookPath() string {
 	return h.GuestHookPath
 }
 
+func (h hypervisor) guestHookTimeout() int32 {
+	if h.GuestHookTimeout <= 0 {
```
Comment (Ji-Xinyou) [inline src/runtime/pkg/katautils/config.go:526]: AFAIC logically this should be `<0` instead of a `<=0` as zero-timeout has its own semantic now.

File: src/runtime/pkg/katautils/config_test.go [hunk 15]
```diff
@@ -1118,6 +1121,21 @@ func TestHypervisorDefaultsGuestHookPath(t *testing.T) {
 	assert.Equal(guestHookPath, testGuestHookPath, "custom guest hook path wrong")
 }
 
+func TestHypervisorDefaultsGuestHookTimeout(t *testing.T) {
+	assert := assert.New(t)
+
+	h := hypervisor{}
+	guestHookTimeout := h.guestHookTimeout()
+	assert.Equal(guestHookTimeout, defaultGuestHookTimeout, "default guest hook path wrong")
```
Comment (Ji-Xinyou) [inline src/runtime/pkg/katautils/config_test.go:1129]: Same here, this test will fail for sure.

File: src/runtime/pkg/oci/utils.go [hunk 16]
```diff
@@ -475,6 +475,16 @@ func addHypervisorConfigOverrides(ocispec specs.Spec, config *vc.SandboxConfig,
 		}
 	}
 
+	if value, ok := ocispec.Annotations[vcAnnotations.GuestHookTimeout]; ok {
+		if value != "" {
+			timeout, err := strconv.ParseInt(value, 10, 32)
+			if err != nil {
+				return err
+			}
+			config.HypervisorConfig.GuestHookTimeout = int32(timeout)
```
Comment (Ji-Xinyou) [inline src/runtime/pkg/oci/utils.go:484]: I just want to note that if you change the message's `int32` to `uint32`, you should notice this kind of places.

File: src/runtime/virtcontainers/hypervisor.go [hunk 17]
```diff
@@ -333,6 +333,10 @@ type HypervisorConfig struct {
 	// GuestHookPath is the path within the VM that will be used for 'drop-in' hooks
 	GuestHookPath string
 
+	// GuestHookTimeout is the number of seconds to wait for a hook execution before aborting it.
+	//It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
+	GuestHookTimeout int32
```
Comment (Ji-Xinyou) [inline src/runtime/virtcontainers/hypervisor.go:338]: ditto

File: src/runtime/virtcontainers/persist/api/config.go [hunk 18]
```diff
@@ -86,6 +86,9 @@ type HypervisorConfig struct {
 	// GuestHookPath is the path within the VM that will be used for 'drop-in' hooks
 	GuestHookPath string
 
+	// GuestHookTimeout is the number of seconds before aborting the hook, If set, it must be greater than zero.
+	GuestHookTimeout int32
```
Comment (Ji-Xinyou) [inline src/runtime/virtcontainers/persist/api/config.go:90]: ditto

**Review by Ji-Xinyou — CHANGES_REQUESTED** [review]
Generally LGTM. Thanks for the nice work @soulseen.
Some bugs and suggestions below.

File: src/libs/protocols/protos/agent.proto (hunk 13 above)
Comment (soulseen) [reply src/libs/protocols/protos/agent.proto:300]: https://github.com/kata-containers/kata-containers/blob/main/src/libs/oci/src/lib.rs#L188
according to the Hook struct defined in Rust, it also used `i32`, so I think it should be consistent.

**Comment by soulseen** [comment]
@Ji-Xinyou please take a look again, thanks a lot!!

File: src/runtime/config/configuration-acrn.toml.in [hunk 19]
```diff
@@ -117,7 +117,11 @@ block_device_driver = "@DEFBLOCKSTORAGEDRIVER_ACRN@"
 # Warnings will be logged if any error is encountered while scanning for hooks,
 # but it will not abort container execution.
 #guest_hook_path = "/usr/share/oci/hooks"
-
+#
+# Guest hook timeout is the number of seconds to wait for a hook execution before aborting it.
+# It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
+#guest_hook_timeout = 10
+#
```
Comment (Ji-Xinyou) [inline src/runtime/config/configuration-acrn.toml.in:124]: This is trivial, but I think you should leave line 124 blank.

File: src/runtime/config/configuration-clh.toml.in [hunk 20]
```diff
@@ -195,6 +195,10 @@ block_device_driver = "virtio-blk"
 # but it will not abort container execution.
 #guest_hook_path = "/usr/share/oci/hooks"
 #
+# Guest hook timeout is the number of seconds to wait for a hook execution before aborting it.
+# It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
+#guest_hook_timeout = 10
+#
```
Comment (Ji-Xinyou) [inline src/runtime/config/configuration-clh.toml.in:201]: ditto

File: src/runtime/config/configuration-fc.toml.in [hunk 21]
```diff
@@ -210,6 +210,10 @@ valid_entropy_sources = @DEFVALIDENTROPYSOURCES@
 # but it will not abort container execution.
 #guest_hook_path = "/usr/share/oci/hooks"
 #
+# Guest hook timeout is the number of seconds to wait for a hook execution before aborting it.
+# It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
+#guest_hook_timeout = 10
+#
```
Comment (Ji-Xinyou) [inline src/runtime/config/configuration-fc.toml.in:216]: ditto

File: src/runtime/config/configuration-qemu.toml.in [hunk 22]
```diff
@@ -391,6 +391,10 @@ valid_entropy_sources = @DEFVALIDENTROPYSOURCES@
 # but it will not abort container execution.
 #guest_hook_path = "/usr/share/oci/hooks"
 #
+# Guest hook timeout is the number of seconds to wait for a hook execution before aborting it.
+# It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
+#guest_hook_timeout = 10
+#
```
Comment (Ji-Xinyou) [inline src/runtime/config/configuration-qemu.toml.in:397]: ditto

File: src/runtime/virtcontainers/hypervisor.go [hunk 23]
```diff
@@ -333,6 +333,10 @@ type HypervisorConfig struct {
 	// GuestHookPath is the path within the VM that will be used for 'drop-in' hooks
 	GuestHookPath string
 
+	// GuestHookTimeout is the number of seconds to wait for a hook execution before aborting it.
+	//It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
```
Comment (Ji-Xinyou) [inline src/runtime/virtcontainers/hypervisor.go:337]: Add a space after the `//`

**Review by Ji-Xinyou — COMMENTED** [review]
Only some formatting stuff.
I personally approve all other parts. Nice work! @soulseen.

**Review by Ji-Xinyou — APPROVED** [review]
LGTM, thanks for the work @soulseen!

**Comment by Ji-Xinyou** [comment]
> This PR modified `src/libs/protocols/protos/agent.proto`, that is, `runtime-rs` should be updated too, and also the `src/libs/kata-types` to keep the synchronization between all components.
> 
> Update agent protocol may be a heavy work, should we pass some additional config info through annotation?
> 
> /cc @kata-containers/architecture-committee
> 
> @soulseen, from the user perspective, they can only set timeout through annotation? I think it should also give user the change to set it by modifying the configuration file to do a node level setup, like:
> 
> https://github.com/kata-containers/kata-containers/blob/33360f1710451293d7fcce301de02a16a1000139/src/runtime/config/configuration-qemu.toml.in#L363

\cc @liubin , I have raised an issue https://github.com/kata-containers/kata-containers/issues/5935 for the integration into runtime-rs.

File: src/runtime/pkg/oci/utils.go [hunk 24]
```diff
@@ -475,6 +475,16 @@ func addHypervisorConfigOverrides(ocispec specs.Spec, config *vc.SandboxConfig,
 		}
 	}
 
+	if value, ok := ocispec.Annotations[vcAnnotations.GuestHookTimeout]; ok {
```
Comment (liubin) [inline src/runtime/pkg/oci/utils.go:478]: This should use the `newAnnotationConfiguration().setInt()`, please refer to the code around here.

**Comment by liubin** [comment]
And also please fix ci error:

> Error: utils.go:208:27: fieldalignment: struct with 272 pointer bytes could be 264 (govet)
type RuntimeConfigOptions struct {
                          ^
make: *** [Makefile:41: static-checks] Error 1

**Comment by soulseen** [comment]
@liubin fixed, please take a review, thanks!

**Comment by liubin** [comment]
@soulseen you need a rebase now.

File: src/agent/src/rpc.rs (hunk 9 above)
Comment (soulseen) [reply src/agent/src/rpc.rs:1212]: > 1. The guest hook is a kata-defined hook. It is not an OCI hook.
> 2. We should allow zero timeout value for those who aren't sure about how long their hooks are expected to execute.
> 
> I'd suggest something like this:
> 
> ```
> Guest hook timeout is the number of seconds to wait for a hook execution before aborting it.
> It must be a non-negative integer. Zero means no timeout. The default value is 10 seconds.
> ```

@Ji-Xinyou @bergwolf I think the `guest hook` here is confusing. actually, it will be passed to OCI Hook struct and the `prestart hook` will be executed after the guest VM start and before the container start.
